Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool.swift

```diff
@@ -121,6 +128,364 @@ enum HTTPConnectionPool {
+        self.key = key
+        self.delegate = delegate
+        self.logger = logger
+
+        self.idleConnectionTimeout = clientConfiguration.connectionPool.idleTimeout
+
+        self._state = StateMachine(
+            eventLoopGroup: eventLoopGroup,
+            idGenerator: idGenerator,
+            maximumConcurrentHTTP1Connections: 8
+        )
+    }
+
+    func executeRequest(_ request: HTTPSchedulableRequest) {
+        let action = self.stateLock.withLock { () -> StateMachine.Action in
+            self._state.executeRequest(.init(request))
+        }
+        self.run(action: action)
+    }
+
+    func shutdown() {
+        let action = self.stateLock.withLock { () -> StateMachine.Action in
+            self._state.shutdown()
+        }
+        self.run(action: action)
+    }
+
+    func run(action: StateMachine.Action) {
+        self.runConnectionAction(action.connection)
+        self.runRequestAction(action.request)
+    }
+
+    func runConnectionAction(_ action: StateMachine.ConnectionAction) {
+        switch action {
+        case .createConnection(let connectionID, let eventLoop):
+            self.createConnection(connectionID, on: eventLoop)
+
+        case .scheduleBackoffTimer(let connectionID, let backoff, on: let eventLoop):
+            self.scheduleConnectionStartBackoffTimer(connectionID, backoff, on: eventLoop)
+
+        case .scheduleTimeoutTimer(let connectionID, on: let eventLoop):
+            self.scheduleIdleTimerForConnection(connectionID, on: eventLoop)
+
+        case .cancelTimeoutTimer(let connectionID):
+            self.cancelIdleTimerForConnection(connectionID)
+
+        case .closeConnection(let connection, isShutdown: let isShutdown):
+            // we are not interested in the close future...
+            _ = connection.close()
+
+            if case .yes(let unclean) = isShutdown {
+                self.delegate.connectionPoolDidShutdown(self, unclean: unclean)
+            }
+
+        case .cleanupConnections(let cleanupContext, isShutdown: let isShutdown):
+            for connection in cleanupContext.close {
+                _ = connection.close()
+            }
+
+            for connection in cleanupContext.cancel {
+                _ = connection.close()
+            }
+
+            for connectionID in cleanupContext.connectBackoff {
+                self.cancelConnectionStartBackoffTimer(connectionID)
+            }
+
+            if case .yes(let unclean) = isShutdown {
+                self.delegate.connectionPoolDidShutdown(self, unclean: unclean)
+            }
+
+        case .none:
+            break
+        }
+    }
+
+    func runRequestAction(_ action: StateMachine.RequestAction) {
+        switch action {
+        case .executeRequest(let request, let connection, cancelTimeout: let cancelTimeout):
+            connection.executeRequest(request.req)
+            if cancelTimeout {
+                self.cancelRequestTimeout(request.id)
```

Comment:
Added a code comment.

Reply:
Let's look at the definition of the request timeouts.

When a request timeout fires, it will tell the state machine about the it having fired. The state machine will then look for the request in its own request queue. If it doesn't find the request in there, it knows that the request was already removed (probably in another thread) before. The action will for this reason be none.

```swift
    private func scheduleRequestTimeout(_ request: Request, on eventLoop: EventLoop) {
        let requestID = request.id
        let scheduled = eventLoop.scheduleTask(deadline: request.connectionDeadline) {
            // The timer has fired. Now we need to do a couple of things:
            //
            // 1. Remove ourselves from the timer dictionary to not leak any data. If our
            //    waiter entry still exists, we need to tell the state machine, that we want
            //    to fail the request.
            let timeoutFired = self.timerLock.withLock {
                self._requestTimer.removeValue(forKey: requestID) != nil
            }

            // 2. If the entry did not exists anymore, we can assume that the request was
            //    scheduled on another connection. The timer still fired anyhow because of a
            //    race. In such a situation we don't need to do anything.
            guard timeoutFired else { return }

            // 3. Tell the state machine about the timeout
            let action = self.stateLock.withLock {
                self._state.timeoutRequest(requestID)
            }

            self.run(action: action)
        }

        self.timerLock.withLockVoid {
            assert(self._requestTimer[requestID] == nil)
            self._requestTimer[requestID] = scheduled
        }

        request.req.requestWasQueued(self)
    }
```